Source/JavaScriptCore/dfg/DFGAbstractHeap.h

```diff
@@ -123,7 +123,7 @@ class AbstractHeap {
         
         Payload(const void* pointer)
             : m_isTop(false)
-            , m_value(bitwise_cast<intptr_t>(pointer))
+            , m_value(reinterpret_cast<int64_t>(pointer))
```

Comment:
I think `static_cast<int64_t>(reinterpret_cast<uintptr_t>(pointer));` maybe better here since
* It is more portable, working properly on both 32-bit and 64-bit systems.
* It has correct semantics, treating pointers as unsigned values, which they are by nature.
* It ensures zero extension, avoiding any risk of sign extension when casting from a smaller pointer to a 64-bit integer.

Reply:
Out of personal interest, can you explain why/how your proposed expression is more portable/correct? I was going off the following clause for reinterpret_cast on cppreference:

> 2) A pointer can be converted to any integral type large enough to hold all values of its type (e.g. to [std::uintptr_t](https://en.cppreference.com/w/cpp/types/integer)).

Which to me reads as if this should work for both 64 and 32 bit platforms since int64_t can hold all values of both 64 bit and 32 bit pointers. Maybe I am missing something?

Reply:
Yes, `reinterpret_cast` performs a bitwise reinterpretation. I think the proposed solution offers improves readability:

Portability: `uintptr_t` adapts to the pointer size on both 32-bit and 64-bit systems, ensuring cross-platform compatibility.
Correct Semantics: Pointers are unsigned, and `uintptr_t` is specifically designed to handle unsigned pointer values.
Safety: While `reinterpret_cast<int64_t>(pointer)` reinterprets the bit pattern without extensions, using `uintptr_t` ensures that the pointer’s value is correctly preserved, avoiding potential truncation or sign-extension issues.